src/inmanta/deploy/scheduler.py

```diff
@@ -0,0 +1,163 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import asyncio
+from collections.abc import Mapping, Set
+from typing import Optional
+
+from inmanta.data.model import ResourceIdStr
+from inmanta.deploy import work
+from inmanta.deploy.state import ModelState, ResourceDetails, ResourceStatus
+
+
+# TODO: name
+# TODO: expand docstring
+class Scheduler:
+    """
+    Scheduler for resource actions. Reads resource state from the database and accepts deploy, dry-run, ... requests from the
+    server. Schedules these requests as tasks according to priorities and, in case of deploy tasks, requires-provides edges.
+    """
+    def __init__(self) -> None:
+        self._state: ModelState = ModelState(version=0)
+        self._work: work.ScheduledWork = work.ScheduledWork(requires=model_state.requires)
+
+        # We uphold two locks to prevent concurrency conflicts between external events (e.g. new version or deploy request)
+        # and the task executor background tasks.
+        #
+        # - lock to block scheduler state access (both model state and scheduled work) during scheduler-wide state updates
+        #   (e.g. trigger deploy). A single lock suffices since all state accesses (both read and write) by the task runners are
+        #   short, synchronous operations (and therefore we wouldn't gain anything by allowing multiple readers).
+        # TODO: this lock may have little value left now. Consider.
+        self._scheduler_lock: asyncio.Lock = asyncio.Lock()
+        # - lock to serialize scheduler state updates (i.e. process new version)
+        self._update_lock: asyncio.Lock = asyncio.Lock()
+
+    def start(self) -> None:
+        # TODO (ticket): read from DB instead
+        pass
+
+    async def deploy(self) -> None:
```

Comment:
this I assume is the external deploy trigger (i.e. current deploy button)

Reply:
External deploy trigger yes, and deploy button is one case, but it can also be the server's deploy timer triggering it (for now).